docs/dev/guidelines/server.rst

```diff
@@ -431,23 +429,17 @@ For more details, please visit the :doc:`./criteria-builder` page.
 22. REST endpoint best practices for authorization
 ==================================================
 
-To reject unauthorized requests as early as possible, Artemis employs a two-step system:
-
-#. ``PreAuthorize`` and ``Enforce`` annotations are responsible for blocking users with wrong or missing authorization roles without querying the database.
-#. The ``AuthorizationCheckService`` is responsible for checking access rights to individual resources by querying the database.
+To reject unauthorized requests as early as possible, Artemis employs two solutions:
 
-Because the first method without database queries is substantially faster, always annotate your REST endpoints with the corresponding annotation. Always use the annotation for the minimum role that has access.
-The following example makes the call only accessible to ADMIN and INSTRUCTOR users:
-
-.. code-block:: java
+#. Implicit pre- and post-authorization annotations operate in two steps:
+    #. ``EnforceRoleInResource`` (e.g. ``EnforceRoleInCourse``) annotations perform pre authorization checks to block users with wrong or missing authorization roles without querying the database.
+    #. If necessary we check for access rights to individual resources within the database via light-weight queries.
+    #. Currently we offer the following annotations: ``EnforceRoleInCourse`` (e.g. ``EnforceAtLeastInstructorInCourse``), and ``EnforceRoleInExercise`` (e.g. ``EnforceAtLeastInstructorInExercise``)
+#. Explicit authorization checks:
+    #. ``PreAuthorize`` and ``Enforce`` annotations are responsible for blocking users with wrong or missing authorization roles without querying the database.
+    #. The ``AuthorizationCheckService`` is responsible for checking access rights to individual resources by querying the database.
```

Comment:
Some change suggestions to make the code more readable (PreAuthorize was removed since it is only used for the EnforceAtLeastRole annotations)

```suggestion
#. Implicit pre- and post-authorization annotations:
    #. ``EnforceRoleInResource`` (e.g. ``EnforceAtLeastInstructorInCourse``) annotations are responsible for blocking users with wrong or missing authorization roles without querying the database.
    #. If necessary, these annotations check for access rights to individual resources within the database via light-weight queries.
    #. Currently we offer the following annotations: ``EnforceRoleInCourse`` and ``EnforceRoleInExercise``
#. Explicit authorization checks (which operate in two steps):
    #. ``EnforceAtLeastRole`` (e.g. ``EnforceAtLeastInstructor``) annotations are responsible for blocking users with wrong or missing authorization roles without querying the database.
    #. The ``AuthorizationCheckService`` is responsible for checking access rights to individual resources by querying the database.
```